### DISCUSSION TITLE
What is the reason for both `getX` and `initialiseX`?

### DISCUSSION THREAD

**Discussion by akauppi** [firebase/firebase-js-sdk#4659] [Q&A]
The first thing a developer faces with the alpha API is how to get access to a subpackage.

Here, we have `getAuth` and `initializeAuth`. Which to use?

![image](https://user-images.githubusercontent.com/1297734/111829191-0b20dc00-88f5-11eb-9ec4-a0e9d55311f0.png)

I've thought that `getAuth` is intended to be the "give me defaults, please" option, but can see three problems:

- defaults are not clearly mentioned in the source comments (see `getPerformance`, below). I need to go find them somewhere... (I don't know, where. [Not in the docs](https://modularfirebase.web.app/reference/performance.getperformance))
- also `initializePerformance` has defaults. The `settings` parameter is optional (see below, again)
- having two names for what seems to be the same thing is adding to the mental footprint of using these packages

```
/**
 * Returns a FirebasePerformance instance for the given app.
 * @param app - The FirebaseApp to use.
 * @public
 */
export declare function getPerformance(app: FirebaseApp): FirebasePerformance;
/**
 * Returns a FirebasePerformance instance for the given app. Can only be called once.
 * @param app - The FirebaseApp to use.
 * @param settings - Optional settings for the Performance instance.
 * @public
 */
export declare function initializePerformance(app: FirebaseApp, settings?: PerformanceSettings): FirebasePerformance;
```

Isn't the `initializePerformance` prototype the only one really needed?

**Comment by Feiyang1** [comment]
Their intended use cases are 
- `get[X]` initializes a service using default options and dependencies. Or just get the existing instance if it has already been created
- `initialize[X]` allows you to customize how a service is created, giving you the full control over your dependencies. It can only be called once for the same app instance

And here are my thoughts to address the problems

- We will update the comment (thus the reference docs) to clearly state what the defaults are
- We can make `settings` required. I'm not too sure about this one because using defaults is a valid choice, but I do see your point that having 2 APIs be able to do the same thing is confusing.
- We can potentially rename `get[X]` to something like `getOrInit[X]WithDefault` to make the intent clear
- We can explore the possibility of getting rid of `get[X]` functions, but then you will be responsible to keep track of the instance you created by `initialize[X]`
